src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/planning/SpaceBookingDayPlanner.kt

```diff
@@ -0,0 +1,147 @@
+   * assign bookings to beds to determine a premise's capacity on a given day.
+   *
+   * We do not currently intend to present the 'booking to room', instead we are only using this planning
+   * to determine the number of free spaces after bookings are planned.
+   *
+   * If we were to present 'booking to room' planning, we may wish to discuss further optimisation
+   * with premises e.g. would they want to plan bookings into the fullest multiple occupancy room possible,
+   * Or is the goal to minimise shared rooms? What's more important, minimising characteristic surplus, or
+   * minimising single bookings into room with the least other beds?
+   *
+   * Planning goals:
+   *
+   * <ol>
+   *   <li>Priority is given to bookings with constraints, those with the most constraints being planned first</li>
+   *   <li>Bookings are placed into rooms with the least surplus characteristics</li>
+   *   <li>Single occupancy bookings are never placed in rooms with other bookings</li>
+   *   <li>If there are multiple competing rooms (after above goals are satisfied), single occupancy bookings will
+   *   be placed into rooms with the least beds (ensuring minimum 'wastage' in these scenarios)</li>
+   * </ol>
+   *
+   * @param beds All beds available on the day being planned (i.e. won't include out of service beds or beds inactive according to start/end date)
+   * @param bookings All required bookings. The code assumes that the premise level characteristics have already been met
+   */
+  fun plan(
+    beds: Set<Bed>,
+    bookings: Set<SpaceBooking>,
+  ): DayPlannerResult {
+    val bedLedger = BedLedger(beds)
+    val planned = mutableListOf<DayPlannerPlannedBooking>()
+    val unplanned = mutableListOf<DayPlannerUnplannedBooking>()
+
+    val sortedBookings = bookings.toList()
+      .sortedByDescending { b -> b.requiredCharacteristics.sumOf { c -> c.weighting } }
+      .toMutableList()
+
+    sortedBookings.forEach { booking ->
+      val requiresSingleRoom = booking.requiredCharacteristics.any { it.singleRoom }
+      val characteristicsExcludingSingleRoom = booking.requiredCharacteristics.filter { !it.singleRoom }.toSet()
+
+      when (
+        val findResult = bedLedger.findBed(
+          characteristics = characteristicsExcludingSingleRoom,
+          requiresSingleRoom = requiresSingleRoom,
+        )
+      ) {
+        is FindBedResult.BedsFound -> {
+          findResult.beds.forEach { bed ->
+            planned.add(
+              DayPlannerPlannedBooking(
+                bed = bed,
+                booking = booking,
+              ),
+            )
+            bedLedger.reserve(bed)
+          }
+        }
+
+        is FindBedResult.BedNotFound -> {
+          unplanned.add(
+            DayPlannerUnplannedBooking(
+              booking = booking,
+            ),
+          )
+        }
+      }
+    }
+
+    return DayPlannerResult(
+      planned = planned.toList(),
+      unplanned = unplanned.toSet(),
```

Comment:
Am I right thinking that the unplanned here will overbooked bookings, where a suitable bed could not be found?

Reply:
Yep that's correct. I was trying to keep the language neutral wrt. planning, but these are what we would present as 'overbooked'